### PR TITLE
Better unfold

### PR DESCRIPTION
**Pull request by alexandresablayrolles**
Re-implemented unfold2d - leads to a 10% improvement on vision_benchmark

### CODE REVIEW

File: opacus/grad_sample/conv.py [hunk 1]
```diff
@@ -1,16 +1,52 @@
...
+    # print(input.shape, input.stride())
+    strides = list(input.stride())
+    strides = strides[:-2] + [
+        W_pad * dilation[0],
+        dilation[1],
+        W_pad * stride[0],
+        stride[1],
+    ]
+    # print("STRIDES", strides)
+    # print("SHAPE", shape + [kernel_size[0], kernel_size[1], H_effective, W_effective])
```
Comment (ffuuugor) [inline opacus/grad_sample/conv.py:42]: 🦖

File: opacus/grad_sample/conv.py [hunk 2]
```diff
@@ -1,16 +1,52 @@
...
+import time
 import torch
 import torch.nn as nn
+import torch.nn.functional as F
 from opacus.utils.tensor_utils import unfold3d
 
 from .utils import register_grad_sampler
 
 
+def unfold2d(
```
Comment (ffuuugor) [inline opacus/grad_sample/conv.py:16]: This is probably better suited in `opacus/utils/tensor_utils.py`, at least that's where our custom `unfold3d` lives

**Review by ffuuugor — COMMENTED** [review]
Great results!
Can you maybe share more detailed data on your vision benchmark? Either here in PR description and any other place you find suitable. I think it's useful to keep track of experiments we do, so that we have a place to link to when future us are in doubt why we  made certain decision.

File: opacus/grad_sample/conv.py [hunk 3]
```diff
@@ -1,16 +1,52 @@
...
+    input,
+    kernel_size: Tuple[int, int],
+    padding: Tuple[int, int],
+    stride: Tuple[int, int],
+    dilation: Tuple[int, int],
+):
+    # print("input shape", input.shape)
```
Comment (karthikprasad) [inline opacus/grad_sample/conv.py:23]: nit: remove

File: opacus/grad_sample/conv.py [hunk 4]
```diff
@@ -1,16 +1,52 @@
...
+    *shape, H, W = input.shape
```
Comment (karthikprasad) [inline opacus/grad_sample/conv.py:24]: *shape?

File: opacus/grad_sample/conv.py [hunk 5]
```diff
@@ -1,16 +1,52 @@
...
+    H_effective = (
+        H + 2 * padding[0] - (kernel_size[0] + (kernel_size[0] - 1) * (dilation[0] - 1))
+    ) // stride[0] + 1
```
Comment (karthikprasad) [inline opacus/grad_sample/conv.py:27]: Me trying to visualise the shape:
giphy

File: opacus/grad_sample/conv.py (hunk 4 above)
Comment (romovpa) [reply opacus/grad_sample/conv.py:24]: Apparently this puts in `shape` all elements but the last two. Nice, didn't know this syntax!
```python
>>> *shape, H, W = (1, 2, 3, 4, 5)
shape = [1, 2, 3]
```

Comment (karthikprasad) [reply opacus/grad_sample/conv.py:24]: TIL as well!

File: opacus/grad_sample/conv.py (hunk 5 above)
Comment (alexandresablayrolles) [reply opacus/grad_sample/conv.py:27]: Should I comment/document?

Comment (ffuuugor) [reply opacus/grad_sample/conv.py:27]: I find tensor shapes as the most useful comments to help understand this stuff, anything beyond is probably too much.

**Comment by alexandresablayrolles** [comment]
## Benchmark unfold

GPU Quadro GP100

python examples/vision_benchmark.py --device cuda --width 224 --height 224

### Before
GPU Memory 13295MiB
Speed: 234 img/s

### After

GPU Memory 12487MiB
Speed: 274.14 img/s

**Review by ffuuugor — APPROVED** [review]
LGTM, thanks for updating!

Before merging, can you pls fix flake errors (https://app.circleci.com/pipelines/github/pytorch/opacus/1293/workflows/d2cc0e20-7cba-4284-90b2-a5c904ab9cbe/jobs/6707) and rebase on the more recent version so that tests are passing

File: opacus/grad_sample/conv.py (hunk 5 above)
Comment (karthikprasad) [reply opacus/grad_sample/conv.py]: +1 commenting with the shapes should be sufficient. If you can easily explain in a line, that works too; but a paragraph is not worth it IMO.

**Pull request merged**